### ISSUE TITLE
Using subscriptions results in crash

### ISSUE THREAD

**Issue by bretthall** [DieselMeister/Terminal.Gui.Elmish#8]
### Description

When using a subscription that dispatches more than once the program will crash after the first time it fires.

### Repro steps

Here's a minimal reproducing program:
```
open System
open Terminal.Gui.Elmish

type Model = {foo: int}

type Msg = 
    | IncFoo

let rec sub dispatch = 
    let rec loop () = 
        async {
            do! Async.Sleep 1000
            dispatch IncFoo
            return! loop ()
        } 
    loop () |> Async.Start

let init () =
    {foo = 0}, Cmd.none //Cmd.ofSub sub

let update msg model = 
    match msg with
    | IncFoo -> {model with foo = model.foo + 1}, Cmd.none

let view model _ = 
    page [
        label [Text (sprintf "foo = %d" model.foo)]
    ]

[<EntryPoint>]
let main argv =
    try 
        Program.mkProgram init update view
        |> Program.withSubscription (fun _ -> Cmd.ofSub sub)
        |> Program.run
        0
    with
    | exc ->
        IO.File.WriteAllText("error.txt", sprintf "%A" exc)   
        printfn "%A" exc
        1 
```

### Expected behavior

The program should run and update the label once a second.

### Actual behavior

The label is updated once and then nothing happens for around 10 seconds and then the program crashes with the following exception:

```
System.ArgumentException: The view that you end with must be balanced
   at Terminal.Gui.Application.End(View view)
   at Terminal.Gui.Application.RunState.Dispose(Boolean disposing)
   at Terminal.Gui.Application.End(RunState runState)
   at Program.main(String[] argv) in /Users/brett/projects/console-ui-test/Program.fs:line 32
```

### Known workarounds

None

### Related information 

* Operating system: OSX 10.14.6 and Win10
* Branch: 1.3
* .NET Runtime, CoreCLR or Mono Version: .net core 2.2
* Performance information, links to performance testing scripts

**Comment by Hardt-Coded** [comment]
I will look into it. Thank you :)

**Comment by DieselMeister** [comment]
Hi.

The Problem ist, that you start your loop with "Async.Start", which results, that the message is dispatched in another thread. This cause the error, in the underlying Terminal.Gui implementation.

You have to use:

Async.StartImmediate

BUT:

currently, you walk into the problem of this Issue:
https://github.com/DieselMeister/Terminal.Gui.Elmish/issues/7

that the message is only once dispatched. I look into it.

**Comment by bretthall** [comment]
After your advice to use `Async.StartImmediate` so that `dispatch` is called in the UI thread I tried wrapping my call to `dispatch` in `Terminal.Gui.Application.Mainloop.Invoke (fun _ -> dispatch msg)` and that did the trick. Every call to dispatch was handled by the UI and there was no crash. 

Might I suggest changing the `dispatch` that is passed to subscriptions to be `fun msg -> Terminal.Gui.Application.Mainloop.Invoke (dispatch msg)`? It seems like one of the main use cases for subscriptions in an elmish app is to integrate messages coming from other threads so making the `dispatch` that is passed to subscriptions thread safe is probably a good idea. 

Thanks for making this library, it should make doing my hands on sessions at OpenF# (https://www.openfsharp.org/speakers/2019/brett-hall/) a lot easier.

**Comment by bretthall** [comment]
Hi-

After your advice I was able to solve the problem by wrapping my call to dispatch in Mainloop.Invoke (that was a little easier than using StartImmediate and having to switch the execution context in the async workflow). 

I didn’t have any issue with the subscription firing only once. Every my async called dispatch the UI was updated appropriately.

Thanks.

-brett

**Comment by Hardt-Coded** [comment]
Sorry, wrong advise. I think I have it already ;)

Mainloop invoke is a way. But as it seems, even that is not necessary.

Give me several minutes and i push a solutions :D

**Comment by Hardt-Coded** [comment]
> Thanks for making this library, it should make doing my hands on sessions at OpenF# (https://www.openfsharp.org/speakers/2019/brett-hall/) a lot easier.

Wow ... I feel honered ... awesome !

**Comment by Hardt-Coded** [comment]
Okay, it take a while. I struggle with the refocus of the selected elements. I have to research a little to find a better solution.

**Comment by bretthall** [comment]
You don’t need to rush on my account. I have a workable solution for what I need to do.

-brett

**Comment by Hardt-Coded** [comment]
I have fixed this. But with that other "flaws" appear. I will fix that. You can download the master branch, if you will and test it.

**Comment by bretthall** [comment]
Tested it out and things are working fine.

-brett

**Comment by Hardt-Coded** [comment]
Hi. So I published a new nuget version. 

https://www.nuget.org/packages/Terminal.Gui.Elmish/0.1.4

feel free to report further issues. :)

**Comment by bretthall** [comment]
The nuget package appears to be broken. After upgrading all the stuff in the `Program` module is gone and the only thing in there is a function called `mkConsoleProgram` that requires me to add a reference to the Fable.Elmish package in order to get the `Elmish.Cmd` type. After adding that package I can create a program, but can't run it since `Program.run` is gone. From looking at the code in github it looks like 0.1.4 should not have these changes. Any idea what's going on?

**Comment by Hardt-Coded** [comment]
Mmmh strange. I will look into it. 




Holen Sie sich Outlook für Android







On Sun, Aug 25, 2019 at 1:14 AM +0200, "Brett Hall" <notifications@github.com> wrote:












The nuget package appears to be broken. After upgrading all the stuff in then Program module is gone and the only thing in there is a function called mkConsoleProgram that requires me to add a reference to the Fable.Elmish package in order to get the Elmish.Cmd type. After adding that package I can create a program, but can't run it since Program.run is gone. Form looking at the code in github it looks like 0.1.4 should not have these changes. Any idea what's going on?

**Comment by Hardt-Coded** [comment]
I can see the behavior, but why? I did not use any Fable.Stuff in there.

**Comment by Hardt-Coded** [comment]
it seems, that maybe I have some old dll in the release folders on my computer, which are old. wait I publish a new package.

**Comment by Hardt-Coded** [comment]
the new package is published and the old is unlisted. Sorry for the inconvenience. version 0.1.5 should work

**Comment by bretthall** [comment]
Great, it seems to be working fine now. Thanks for the speedy fixes.

-brett

**Comment by drk-mtr** [comment]
Is anyone able to advise on how to get this working? This is still broken for me when I try to run the Elmish.Console.Elements example.

**Comment by Hardt-Coded** [comment]
I am on it. Sorry, had to work a lot for the last view weeks.

**Comment by drk-mtr** [comment]
Thanks for the reply, and please don't apologise - I just wish I could do more to help! Thanks for sharing the library, it's great.

**Comment by Hardt-Coded** [comment]
@drk-mtr  should have been fixed. :)

**Comment by drk-mtr** [comment]
Just putting a note here in case anyone stumbles across it... this was fixed for me on mac in release 0.1.6 :)
